### PR TITLE
Feature/source add function

### PR DESCRIPTION
**Pull request by creisle**
Improvements
- Adds a helper function for some commonly re-used code. Also ensures that this default function will try to not create multiple sources with the same name
- Improve performance on reloading DO with caching option
- Improve load robustness by retrying GraphKB requests if there is a server side error

### CODE REVIEW

**Comment by creisle** [comment]
Sorry, pushed changes to the wrong branch

**Comment by creisle** [comment]
Let me know if you'd prefer I remake this PR, otherwise I'm going to leave them grouped since it is simpler. Has the function refactor and the caching adding for DO

File: src/ontology/index.js [hunk 1]
```diff
@@ -154,13 +154,7 @@ const uploadFromJSON = async ({ data, conn }) => {
...
     try {
         await Promise.all(Object.entries(sources).map(async ([sourceKey, sourceDefn]) => {
-            const sourceRID = rid(await conn.addRecord({
-                content: sourceDefn,
-                existsOk: true,
-                fetchConditions: { name: sourceDefn.name },
-                target: 'Source',
-                upsert: true,
-            }));
+            const sourceRID = rid(await conn.addSource(sourceDefn));
```
Comment (elewis2) [inline src/ontology/index.js:157]: fetchConditions and upsert aren't provided by addRecord - are they not needed here?

Comment (elewis2) [reply src/ontology/index.js:157]: sorry, just upsert

File: src/hgnc/index.js [hunk 2]
```diff
@@ -227,13 +227,7 @@ const fetchAndLoadBySymbol = async ({
...
         hgnc = CACHE.SOURCE;
     } else {
-        hgnc = await conn.addRecord({
-            content: SOURCE_DEFN,
-            existsOk: true,
-            fetchConditions: { name: SOURCE_DEFN.name },
-            fetchExisting: true,
-            target: 'Source',
-        });
+        hgnc = await conn.addSource(SOURCE_DEFN);
```
Comment (elewis2) [inline src/hgnc/index.js:230]: fetchExisting isn't provided by addRecord, is it not actually needed here?

File: src/graphkb.js [hunk 3]
```diff
@@ -230,25 +241,31 @@ class ApiConnection {
      * @param {string} opt.uri the uri target
      * @param {object} opt.body the request body
      * @param {object} opt.qs the query parameters
+     * @param {nuumber} opt.retries number of times to retry the request
```
Comment (elewis2) [inline src/graphkb.js:244]: nuumber -> number

File: src/hgnc/index.js (hunk 2 above)
Comment (creisle) [reply src/hgnc/index.js:230]: fetchExisting can be left off b/c it is default true

File: src/ontology/index.js (hunk 1 above)
Comment (creisle) [reply src/ontology/index.js:157]: good catch! i think upsert must have been lost in a merge or something. I will re-add it

File: src/graphkb.js (hunk 3 above)
Comment (creisle) [reply src/graphkb.js]: fixed

File: src/ontology/index.js (hunk 1 above)
Comment (creisle) [reply src/ontology/index.js:157]: fixed

**Review by elewis2 — APPROVED** [review]

**Pull request merged**